File: backend/spring-boot/src/main/java/org/bootstrapbugz/backend/auth/security/UserDetailsServiceImpl.java
package org.bootstrapbugz.backend.auth.security;

import org.bootstrapbugz.backend.shared.error.exception.UnauthorizedException;
import org.bootstrapbugz.backend.shared.message.service.MessageService;
import org.bootstrapbugz.backend.user.repository.UserRepository;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class UserDetailsServiceImpl implements UserDetailsService {
  private final UserRepository userRepository;
  private final MessageService messageService;

  public UserDetailsServiceImpl(UserRepository userRepository, MessageService messageService) {
    this.userRepository = userRepository;
    this.messageService = messageService;
  }

  @Transactional
  public UserDetails loadUserByUserId(Long userId) {
    return userRepository
        .findById(userId)
        .map(UserPrincipal::create)
        .orElseThrow(() -> new UnauthorizedException(messageService.getMessage("auth.unauthorized")));
  }

  @Override
  @Transactional
  public UserDetails loadUserByUsername(String username) {
    return userRepository
        .findByUsernameOrEmail(username, username)
        .map(UserPrincipal::create)
        .orElseThrow(() -> new UnauthorizedException(messageService.getMessage("auth.unauthorized")));
  }
}

File: backend/spring-boot-kotlin/src/main/kotlin/org/bootstrapbugz/api/auth/security/UserDetailsServiceImpl.kt
package org.bootstrapbugz.api.auth.security

import org.bootstrapbugz.api.shared.error.exception.UnauthorizedException
import org.bootstrapbugz.api.shared.message.service.MessageService
import org.bootstrapbugz.api.user.repository.UserRepository
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UserDetailsServiceImpl(
    private val userRepository: UserRepository,
    private val messageService: MessageService
) : UserDetailsService {

    @Transactional
    fun loadUserByUserId(userId: Long): UserDetails =
        userRepository.findById(userId)
            .map(UserPrincipal.Companion::create)
            .orElseThrow { UnauthorizedException(messageService.getMessage("token.invalid")) }

    @Transactional
    override fun loadUserByUsername(username: String): UserDetails =
        userRepository.findByUsernameOrEmail(username, username)
            .map(UserPrincipal.Companion::create)
            .orElseThrow { UnauthorizedException(messageService.getMessage("token.invalid")) }
}
